sql/sql_acl.cc

```diff
@@ -13736,45 +13739,37 @@ static ulong parse_client_handshake_packet(MPVIO_EXT *mpvio,
   THD *thd= mpvio->auth_info.thd;
   NET *net= &thd->net;
   char *end;
+  DBUG_ENTER("parse_client_handshake_packet");
   DBUG_ASSERT(mpvio->status == MPVIO_EXT::FAILURE);
+  DBUG_PRINT("info", ("parse_client_handshake STARTING: vio_type=%s",
+                      safe_vio_type_name(thd->net.vio)));
 
   if (pkt_len < MIN_HANDSHAKE_SIZE)
-    return packet_error;
+    DBUG_RETURN(packet_error);
 
   /*
     Protocol buffer is guaranteed to always end with \0. (see my_net_read())
     As the code below depends on this, lets check that.
   */
   DBUG_ASSERT(net->read_pos[pkt_len] == 0);
 
-  ulonglong client_capabilities= uint2korr(net->read_pos);
-  compile_time_assert(sizeof(client_capabilities) >= 8);
-  if (client_capabilities & CLIENT_PROTOCOL_41)
+  ushort first_two_bytes_of_client_capabilities= uint2korr(net->read_pos);
+  if (first_two_bytes_of_client_capabilities & CLIENT_SSL)
   {
-    if (pkt_len < 32)
-      return packet_error;
-    client_capabilities|= ((ulong) uint2korr(net->read_pos+2)) << 16;
-    if (!(client_capabilities & CLIENT_MYSQL))
-    {
-      // it is client with mariadb extensions
-      ulonglong ext_client_capabilities=
-        (((ulonglong)uint4korr(net->read_pos + 28)) << 32);
-      client_capabilities|= ext_client_capabilities;
-    }
-  }
-
-  /* Disable those bits which are not supported by the client. */
-  compile_time_assert(sizeof(thd->client_capabilities) >= 8);
-  thd->client_capabilities&= client_capabilities;
```

Comment:
The diff makes this look more complicated than it really is.

All that is really changing is that these few lines move from _before_ the TLS handshake to _after_ the TLS handshake.

These line tells the server to ***read capability bits sent by the client and to trust them***: the server alters its view of how the connection should work (`thd->client_capabilities&= `) based on them.

- If the server reads these bits from the pre-TLS handshake plaintext version of the login request packet (https://mariadb.com/kb/en/connection/#sslrequest-packet), it **forces the client to send this juicy identifying information in the clear**.
- Making the server instead read these bits from the post-TLS handshake version of the login request packet (https://mariadb.com/kb/en/connection/#handshake-response-packet), it allows the client to *only* send its complete capability bits (and other identifying information) over the TLS-secured connection.
   - … and the corresponding Connector/C change allows the client to do just that: ***TBD***